### ISSUE TITLE
shouldUpdate purpose

### ISSUE THREAD

**Issue by Frikki** [dekujs/deku#186]
I am a bit confused regarding how you intended `shouldUpdate()` to work. As I have understood, returning **false** would skip the `render()` method, but I wasn’t aware that it completely would skip setting the next state, which seems to be the case in v0.4.5.

**Comment by anthonyshort** [comment]
It's designed to just skip calling `render` (so the virtual nodes don't need to be created and diffed) but it will still check/update the child components. 

We could potentially move the `commit` call higher and set it anyway: https://github.com/dekujs/deku/blob/master/lib/render.js#L346

Honestly we've never ended up using `shouldUpdate`. It would be useful if you were using immutable data, but you could get a similar thing by just memoizing some render functions, because it will skip the diffing if the same node is returned: https://github.com/dekujs/deku/blob/master/lib/render.js#L354

**Comment by Frikki** [comment]
For example, `MyComponent` has an attached, say, `onMouseDown` event handler which sets state, but `MyComponent` shouldn’t render; thus, in the `shouldUpdate()` method, the `nextState` argument is examined and compared to `component.state` to return **false** on a given condition, causing render to be skipped. However, it is important that the component state is actually changed. This doesn’t seem to happen if `shouldUpdate()` returns **false**, i.e., the `nextState` is not applied.

The above scenario can cause problems when another event updates state to cause a render, but it is dependent on the previous event’s state change. Such a scenario will force users to add further private properties to keep track of the component’s state, which would be a poor practice.

**Comment by anthonyshort** [comment]
> onMouseDown event handler which sets state, but MyComponent shouldn’t render

When would you change the `state` without wanting a re-render? Are you storing things in state that related to rendering?

**Comment by Frikki** [comment]
> When would you change the state without wanting a re-render?

For example, a `isPointerPressInProgress` state which doesn’t affect render but only events.

> Are you storing things in state that related to rendering?

Yes, state is meant to hold, well, component state; both for rendering and functional purposes.

**Comment by anthonyshort** [comment]
I think we can just change this line:

```js
if (!shouldUpdate(entity)) return updateChildren(entityId)
```

to this:

```js
if (!shouldUpdate(entity)) {
  commit(entity)
  return updateChildren(entityId)
}
```

`shouldUpdate` seems like it should be `shouldRender` since it's actually performing the update just not rendering.

**Comment by anthonyshort** [comment]
You could also store that state outside of the component:

```js
isPointerPressInProgress[component.id] = true
```

that would be better if we were only talking about using `state` only for rendering state (which isn't really clear).

**Comment by Frikki** [comment]
> You could also store that state outside of the component

Which I would find a bad idea, as it is related to the component. It could very well be part of a render in a different context, i.e., in a different combination of state.

> shouldUpdate seems like it should be shouldRender since it's actually performing the update just not rendering.

Agreed.

**Comment by anthonyshort** [comment]
I think we'll just get that change in there to still commit the changes to state, but just skip rendering/diffing. I'd say that was a bug and wasn't working as expected.

**Comment by Frikki** [comment]
If I recall correctly, FB’s React does exactly that, but I am not a 100 percent.

**Comment by anthonyshort** [comment]
> Invoked before rendering when new props or state are being received. This method is not called for the initial render or when forceUpdate is used.
>
> Use this as an opportunity to return false when you're certain that the **transition to the new props and state will not require a component update**.

Sounds like it still applies the state/props but just doesn't render. The word 'update' is probably the wrong word to use everywhere, even in Deku.

**Comment by Frikki** [comment]
The FB React method is called `shouldComponentUpdate`, 

> which is triggered before the re-rendering process starts (virtual DOM comparison and possible eventual DOM reconciliation), giving the developer the ability to short circuit this process.

See https://facebook.github.io/react/docs/advanced-performance.html

So it is about whether a *new* virtual DOM should be generated or not for the component (guess that explains the _update_ part.) New props and state are, however, applied.

**Comment by anthonyshort** [comment]
Sweet, I just merged this in, I'll get 0.4.6 out now!
